### PR TITLE
#35 feat: Exhaust APIs

### PR DESCRIPTION
**Pull request by pveleneni**
Updates -
* Exhaust API code has been added for accessing Backup files from Cloud providers
* Only AWS Support is included as of now
* Test cases have been added accordingly and the coverage is at 92%
* The lib/client-cloud-service are taken from Sunbird cloud-services SDK

### CODE REVIEW

File: src/configs/Config.ts [hunk 1]
```diff
@@ -76,5 +76,11 @@ export const config = {
       "primary_key": "id",
       "references": []
     }
-  }
-}
\ No newline at end of file
+  },
+  "exhaust_provider": process.env.exhaust_provider || "aws", // Supported providers - AWS, GCP, Azure
```
Comment (manjudr) [inline src/configs/Config.ts:80]: Instead of exhaust_provider can we rename it as cloud_storage_provider? or csp_provider?

File: src/configs/Config.ts [hunk 2]
```diff
@@ -76,5 +76,11 @@ export const config = {
...
+  "exhaust_region": process.env.exhaust_region || "", // Region for the cloud provider storage
```
Comment (manjudr) [inline src/configs/Config.ts:81]: Same here as well

File: src/configs/Config.ts [hunk 3]
```diff
@@ -76,5 +76,11 @@ export const config = {
...
+  "exhaust_config": process.env.exhaust_config ? JSON.parse(process.env.exhaust_config) : {}, // Respective credentials object for cloud provider
+  "label_container": process.env.label_container || "", // Storage container/bucket name
```
Comment (manjudr) [inline src/configs/Config.ts:83]: can we rename it as either "container" or "bucket_name"?

File: src/configs/Config.ts [hunk 4]
```diff
@@ -76,5 +76,11 @@ export const config = {
...
+  "label_container_prefix": process.env.label_container_prefix || "", // Path to the folder inside container/bucket. Empty if data at root level
+  "storage_url_expiry": process.env.storage_url_expiry ? parseInt(process.env.storage_url_expiry) : 3600, // in seconds, Default 1hr of expiry for Signed URLs.
```
Comment (manjudr) [inline src/configs/Config.ts:85]: I would suggest you to define one exhaust object and list all the properties which are required for the exhaust API.

File: src/configs/QueryRules.ts [hunk 5]
```diff
@@ -53,3 +54,9 @@ export const queryRules = {
     }
   ]
 }
+
+export const exhaustRules = {
+  "common": {
```
Comment (manjudr) [inline src/configs/QueryRules.ts:59]: Not required @pveleneni  - We can just define the exhaust config object and define all the default values

File: src/lib/client-cloud-services/AWSStorageService.js [hunk 6]
```diff
@@ -0,0 +1,475 @@
+/**
+ * @file        - AWS Storage Service
+ * @exports     - `AWSStorageService`
+ * @since       - 5.0.1
+ * @version     - 1.0.0
+ * @implements  - BaseStorageService
+ * @see {@link https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html | X-Amz-Credential}
+ * @see {@link https://docs.aws.amazon.com/directconnect/latest/APIReference/CommonParameters.html#CommonParameters-X-Amz-Credential | X-Amz-Credential}
+ */
```
Comment (manjudr) [inline src/lib/client-cloud-services/AWSStorageService.js:9]: TODO  - We can publish this cloud service as npm module and use it across all the repositories

File: src/models/DatasetModels.ts [hunk 7]
```diff
@@ -16,6 +16,10 @@ export interface IValidator {
     validate(data: any, id?: string): ValidationStatus | Promise<ValidationStatus>;
 }
 
+export interface QValidator extends IValidator {
```
Comment (manjudr) [inline src/models/DatasetModels.ts:19]: We have already an `IValidator` right @pveleneni  - Can we re use that ?

File: src/services/ClientCloudService.ts [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    private client: any
+    private storage: any
+    private config: any
+    private momentFormat: string;
+    constructor(cloudProvider: string, config?: any) {
+        this.cloudProvider = cloudProvider
+        this.client = CloudService.init(this.cloudProvider)
+        this.config = config
+        this.storage = new this.client(this.config)
+        this.momentFormat = "YYYY-MM-DD HH:MI:SS";
```
Comment (manjudr) [inline src/services/ClientCloudService.ts:24]: `YYYY-MM-DD` Isn't this sufficient? Users will not provide hours and mins.

File: src/services/IngestorService.ts [hunk 9]
```diff
@@ -40,15 +40,29 @@ export class IngestorService {
         throw constants.EMPTY_DATASET_ID
     }
 
+    public async getDatasetConfig(datasetId: string) {
+        let datasetConfigList = globalCache.get("dataset-config");
```
Comment (manjudr) [inline src/services/IngestorService.ts:44]: Why this change here?

**Review by manjudr — CHANGES_REQUESTED** [review]

File: src/services/IngestorService.ts (hunk 9 above)
Comment (pveleneni) [reply src/services/IngestorService.ts:44]: Have refactored to cache the datasets, as the old function was querying the db on each request

File: src/models/DatasetModels.ts (hunk 7 above)
Comment (pveleneni) [reply src/models/DatasetModels.ts:19]: The IValidator allows only 1 method called validate, which is used in both Query Validator and Request validator. Need an additional method to validate query params, so added the interface. 

If not, have add params and check with an if condition for the type of validation to perform

**Comment by manjudr** [comment]
The changes looks fine
